feat: Update arkworks to latest git versions with GR1CS migration

flyingnobita opened the pull request:
## Description
Updates arkworks dependencies to their latest git versions to support GR1CS migration and align with the latest arkworks ecosystem changes.

## Changes
- Updated `ark-groth16` to latest git version (rev: `b3b4a15`)
- Updated all related arkworks dependencies to compatible git versions:
  - `ark-ff`, `ark-ec`, `ark-serialize`, `ark-poly` from arkworks-rs/algebra
  - `ark-relations`, `ark-snark` from arkworks-rs/snark
  - `ark-poly-commit` from arkworks-rs/poly-commit
  - Curve crates (`ark-bn254`, `ark-grumpkin`, `ark-pallas`, `ark-vesta`, `ark-mnt4-298`, `ark-mnt6-298`)
- Migrated from R1CS to GR1CS constraint system where applicable
- Updated rust-toolchain to `1.88.0`

## Temporary Dependencies
Using temporary forks for compatibility:
- `ark-crypto-primitives`: Using fork at [flyingnobita/crypto-primitives](https://github.com/flyingnobita/crypto-primitives) (rev: `f559264`)
- `ark-r1cs-std`: Using fork at [flyingnobita/r1cs-std_yelhousni](https://github.com/flyingnobita/r1cs-std_yelhousni/tree/perf/sw-updated) (rev: `b4bab0c`, "perf/sw-updated" branch)
 - `ark-circom`: Using fork at [dmpierre/circom-compat](https://github.com/dmpierre/circom-compat) (rev: `0dfd773`) to resolve wasm target compilation issues

These forks contain necessary updates for GR1CS compatibility and will be replaced once upstream PRs are merged.

## Testing
- [x] Tests pass locally on Macbook/Apple silicon and Linux/Intel with `cargo test --release --workspace --no-default-features --features default,light-test`
- [x] Examples compile and run successfully
- [x] CI/CD tests manual run passed
  - [x] WASM targets build successfully
    - `cargo build -p experimental-frontends --no-default-features --target wasm32-unknown-unknown --features "wasm, parallel"`
    - `cargo build -p experimental-frontends --no-default-features --target wasm32-wasip1 --features "wasm, parallel"`

## Known Issue:
~~- wasmer 6.1.0-rc.2 has linking issues on macOS ARM64 (undefined libunwind symbols)~~
~~- Workarounds:~~
    ~~- Exclude experimental-frontends from tests on macOS~~
    ~~- Run tests in Linux/Intel environment~~
    ~~- Note: `wasmer 4.4.0` (main) works, but not  on  `wasmer-6.0.1` (dev) and `wasmer-6.1.0-rc.2` (my PR and git version of ark-circom)~~ **RESOLVED**

#### Memory intensive tests involving `decider_eth` do not pass with the `--package` flag 
```shell
cargo test decider_eth --release -p folding-schemes 

test folding::hypernova::decider_eth::tests::test_decider ... FAILED
test folding::hypernova::decider_eth::tests::test_decider_serialization ... FAILED
test folding::hypernova::decider_eth_circuit::tests::test_decider_circuit ... FAILED
test folding::nova::decider_eth::tests::test_decider ... FAILED
test folding::nova::decider_eth::tests::test_decider_serialization ... FAILED
test folding::nova::decider_eth_circuit::tests::test_decider_circuit ... FAILED
test folding::protogalaxy::decider_eth::tests::test_decider ... error: test failed, to rerun pass `-p folding-schemes --lib`

Caused by:
  process didn't exit successfully: `/Users/flyingnobita/Data/Projects/ProgCrypto/ZK/sonobe/target/release/deps/folding_schemes-488098cdcfeb6458` (signal: 9, SIGKILL: kill)
```
  - but it will pass with:
```shell
cargo test decider_eth --release
# or
cargo test decider_eth --release --workspace --exclude experimental-frontends --lib

running 10 tests
test folding::hypernova::decider_eth::tests::test_decider ... ok
test folding::hypernova::decider_eth::tests::test_decider_serialization ... ok
test folding::hypernova::decider_eth_circuit::tests::test_decider_circuit ... ok
test folding::hypernova::decider_eth_circuit::tests::test_lcccs_checker_gadget ... ok
test folding::nova::decider_eth::tests::test_decider ... ok
test folding::nova::decider_eth::tests::test_decider_serialization ... ok
test folding::nova::decider_eth_circuit::tests::test_decider_circuit ... ok
test folding::protogalaxy::decider_eth::tests::test_decider ... ok
test folding::protogalaxy::decider_eth::tests::test_decider_serialization ... ok
test folding::protogalaxy::decider_eth_circuit::tests::test_decider_circuit ... ok
```

## Breaking Changes
None expected for library users. Internal constraint system representations have been updated from R1CS to GR1CS.

## Notes
- This update is necessary to stay aligned with the latest arkworks ecosystem
- The GR1CS (Generalized R1CS) system provides more flexibility for constraint representations
- All existing functionality is preserved

winderica committed:
Add `CycleFoldCommittedInstanceVar::new_incoming_from_components` to simplify the construction of incoming CF instances

winderica committed:
RLC for ProtoGalaxy can now be computed using a single `CycleFoldCircuit`

winderica committed:
Absorb `pp_hash` when initializing sponge or transcript so we no longer need to absorb it later

winderica committed:
Clarify comments and remove redundant code

winderica committed:
Fix Pedersen circuit with inputs of odd length

winderica committed:
Bump rust version

winderica committed:
Fix performance regression of Pedersen circuit

winderica committed:
Attempt to fix wasm target

winderica committed:
Fix features for `getrandom`

winderica committed:
Fix wasmer stuff

winderica committed:
Fix CI

flyingnobita committed:
feat: Update arkworks to latest git versions with GR1CS migration

Dependencies Updated:
- arkworks libraries: Updated to latest git versions via patch.crates-io
  - Note: Pinning to specific revisions caused version conflicts across the arkworks ecosystem
- ark-r1cs-std: Using fork flyingnobita/r1cs-std_yelhousni (rev: b4bab0c) for GR1CS compatibility
- ark-crypto-primitives: Using fork flyingnobita/crypto-primitives (rev: f559264) for GR1CS compatibility
- wasmer: Updated to 6.1.0-rc.2 with sys feature for ark-circom compatibility
- Rust toolchain: Updated to 1.85.0 (required by transitive dependency base64ct v1.8.0 via acvm/noir)

Major API Changes:
- Migrated from R1CS to GR1CS throughout the codebase
  - Changed all R1CSVar imports to GR1CSVar
  - Updated ark_relations::r1cs to ark_relations::gr1cs
- Updated constraint enforcement API
  - Changed enforce_constraint to enforce_r1cs_constraint in experimental-frontends
  - Wrapped linear combinations in closures
- Fixed instance_assignment from field access to method call

Known Issue:
- wasmer 6.1.0-rc.2 has linking issues on macOS ARM64 (undefined libunwind symbols)
  - Workarounds:
    - Exclude experimental-frontends from tests on macOS
    - Run tests in Linux/Intel environment
    - Note: wasmer 6.0.1 works but would require forking ark-circom

winderica commented:
Awesome work, thank you so much for your efforts!

For the `wasmer` issue, can you try to rebase this PR onto [`dev`](https://github.com/privacy-scaling-explorations/sonobe/tree/dev) (this is where the latest development goes, as I don't have access to `main`) and see if it works?

flyingnobita commented:
Thanks!

Yeah okay I'll try that. I'll also take a look at the remaining outstanding failures.

flyingnobita committed:
feat: Complete GR1CS migration and rebase onto dev branch

- Migrate all constraint systems from R1CS to GR1CS across all folding schemes

flyingnobita committed:
refactor: Fix cargo fmt

flyingnobita committed:
fix: Update examples to use GR1CS instead of R1CS

- Replace R1CSVar with GR1CSVar in test modules
- Update ConstraintSystem imports from r1cs to gr1cs
- Fix formatting in experimental-frontends witness calculator

flyingnobita commented:
@winderica I have rebased all the changes to `dev`, and also migrated the examples to use GR1CS. I've also fixed most of the other CI tests. The 3 tests (1 clippy and 2 build targets) related to the wasm targets have some errors which I think require a bit more discussion before fixing. Please take a look and let me know what you think!

flyingnobita committed:
chore: Remove unnecessary features from wasmer dependency to be inline with ark-circom dependency

flyingnobita commented:
After more investigation, it turns out that the `wasmer` problems was related to my Mac using a different clang because of the nix environment that it was built in (nix's `clang v11` vs Apple's `clang v17`). Changing it to use Apple's clang resolves this:

```shell
export CARGO_TARGET_AARCH64_APPLE_DARWIN_LINKER=/usr/bin/clang

# Test as usual
cargo test --release --package experimental-frontends 
```

flyingnobita committed:
fix: Update ark-circom dependency to resolve wasm target compilation

Replace custom ark-circom fork with dmpierre's fork to fix compilation
issues when building experimental-frontends for wasm targets.

dmpierre committed:
ci: trigger

flyingnobita committed:
fix: Fix cargo fmt

winderica reviewed (approved):
Thank you @flyingnobita for investigating and fixing the building issue!

Regarding the failure of memory intensive tests, I think it can be addressed by using [my fork](https://github.com/winderica/r1cs-std/tree/sw-fix), and I will handle it in a separate PR. This PR itself already looks great.

Thank you again for your time and efforts!

flyingnobita commented:
@winderica @dmpierre Thank you for the review and merge! I really appreciate the speed!  🚀🚀
